### ISSUE TITLE
[BUG] `Multinomial.sample` hangs for specific parameter settings and seed

### ISSUE THREAD

**Issue by hylkedonker** [tensorflow/probability#1684]
Hi,

I came across a weird bug where, for a very particular set of parameters of the `Multinomial` distribution, the `sample` function keeps running and doesn't return a sample. I've also encountered circumstances where `Multinomial.sample` does return a sample, but is extremely slow (I haven't managed to track down the specific settings for those cases).

**Example**
Here is a minimal code snippet to reproduce the problem (the file 'probs.npy' I have attached in the zip [probs.zip](https://github.com/tensorflow/probability/files/10439031/probs.zip))
```python
import jax.numpy as jnp
import numpy as np
from tensorflow_probability.substrates import jax as tfp
from sklearn import datasets

tfd = tfp.distributions

digits = datasets.load_digits()
n_samples = len(digits.images)
data = digits.images.reshape((n_samples, -1))

# Parameters for multinomial.
total_count = jnp.array(data[:100], dtype=jnp.float32)
seed = jnp.array([2257234201, 3625399095], dtype=jnp.uint32)

# The file probs.npy can be found in the attached file "probs.zip". 
probs = jnp.load('probs.npy')

# Just to convince you: the `probs` are actual probabilities.
# 0 <= p <= 1.
assert np.all((probs >= 0) & (probs <= 1))
# The probabilities are normalised.
np.testing.assert_array_almost_equal(jnp.sum(probs, axis=-1), 1)

# The following statement hangs.
x_samples = tfd.Multinomial(total_count=total_count, probs=probs).sample(seed=seed)
```
Strangely enough, the problem depends on the value of `seed`. I haven't been able to reduce the number of parameters in  `probs` and `total_count`, so it may be that the size of the arrays is also a factor.

**Environment**
I am running Python 3.10.6 on Ubuntu 22.04.1 LTS with the following package versions:
```
tensorflow==2.9.1
tensorflow-probability==0.19.0
jax==0.4.1
jaxlib==0.4.1
```

Any ideas how to work around this problem? 

Thanks,

Hylke

**Comment by csuter** [comment]
Are you sure tfd.Multinomial is what you want here? Not tfd.Categorical?

Multinomial sampling is based on a batched rejection sampler that, roughly speaking "keeps trying until all the items in the batch are finished". If you pass in lots of batches (here it looks like num images times num pixels?), then it's going to take as long as the slowest batch point. You may get a speed-up by jax.jit'ing the sampler, but I'm curious about what you're actually trying to do here. As written this says "for each image, and each pixel within an image, use the pixel's value as the "number of trials" (`n`) of a [Multinomial distribution](https://en.wikipedia.org/wiki/Multinomial_distribution), sampling from a number of classes (digits) `k` with given probabilities". It's surprising to me that you'd use the pixel value as number of trials. Various sources conflate what we call Multinomial and Categorical distributions in TFP, which leads to my initial question.

**Comment by csuter** [comment]
Just to add to that, I *do* feel like i understand what could be meant using Categorical here:

```
tfd.Categorical(probs=probs).sample(seed=...)
```

Since probs has shape `[num images, num pixels, num classes]`, this would sample an image class for each pixel in each image -- basically telling something like "what each pixel thinks the overal image might be". This makes at least some sense to me, but of course I have no idea what you're really after! Happy to learn more and try to help.

**Comment by csuter** [comment]
if i jit the sampler i get a result in a few seconds:

```
>>> smp = jax.jit(lambda seed: tfd.Multinomial(total_count=total_count, probs=probs).sample(seed=seed))
>>> smp(seed=jax.random.PRNGKey(0)).shape
(100, 64, 10)
```

**Comment by hylkedonker** [comment]
I Christopher,

Thanks for your prompt response. 
I'Il try jitting the function as you suggested. My original code base had several jits, but I removed most of them for debugging purposes. I'Il get back to this.

As for the `Multinomial`: this was in fact intentional. :-)  
Without going in too much detail, I am modelling the  pixels in the images as Poisson distribution using matrix factorisation (with num classes). We rely on the identity
```math
x^{(i)}_{j} = \sum_k x^{(i)}_{jk}; x^{(i)}_{jk} \sim \mathrm{Poisson}(\lambda^{(i)}_{jk})\\
\leftrightarrow \\
x^{(i)}_{jk} \sim \mathrm{Multinomial}[x^{(i)}_{j}, \lambda^{(i)}_{jk} / (\sum_k \lambda^{(i)}_{jk})] ; x^{(i)}_{j} \sim \mathrm{Poisson}(\sum_k \lambda^{(i)}_{jk})
```
for some steps in our Gibbs sampling algorithm. Hence the pixel values ( $x^{(i)}_j$ ) in the multinomial.

I'Il let you know if this simple `jit` is all I needed. Thanks again.

Hylke

**Comment by csuter** [comment]
I see. Neat! Let us know how you fare😊

**Comment by hylkedonker** [comment]
Hi,

I've tested your suggestion. While the `jit`ed sampler works for `seed = jax.random.PRNGKey(0)`, it does not work for `seed = jnp.array([2257234201, 3625399095], dtype=jnp.uint32)`. (Even if I do a warm-up with `smp(seed=jax.random.PRNGKey(0))`).
Any thoughts?

Thanks,

Hylke

**Comment by csuter** [comment]
Interesting. Taking a look.

**Comment by csuter** [comment]
The issue is [here](https://github.com/tensorflow/probability/blob/5936211259e44297283c8fbd3a77e9bf125cd761/tensorflow_probability/python/distributions/multinomial.py#L393). The multinomial sampler works by successively sampling binomials for each class. If the probability mass is exhausted before reaching the last class (ie, some non-zero number of latter classes have zero probability), this line evaluates `0. / 0.`, nan'ing. The underlying binomial sampler, as mentioned above, is a rejection sampler that loops until finished. Nan's are causing an infinite loop there. We can probably make the RS loop logic more robust, but an easy fix here is to use `divide_no_nan` when deriving the binomial probs. I confirmed this change enables the sampler to finish.

I guess the seed dependence comes because we have to get extra unlucky, finding ourselves at the later (zero-prob) classes with more trials still to be run, and perhaps this is just very uncommon, since the trials tend to have been all spent in the earlier non-zero prob classes. Just a guess. I'll put a fix together, which should hit the tfp-nightly package tomorrow.

Thanks for catching and raising this!

**Comment by chrism0dwk** [comment]
@csuter just a reminder -- when we were sorting this out in 2020, I think we identified the issue and did exactly as you suggest above.  Works well for my pre-merge code, but for some reason it didn't make it into the TFP repo.

**Comment by csuter** [comment]
@chrism0dwk Oof...well, it's en route now (in internal code review)!

**Comment by hylkedonker** [comment]
Great, thanks so much guys!

**Comment by csuter** [comment]
Ok the fix should go out in the nightly. Please let us know if you have any more issues. Thanks again!
